drivers/can/can_mcan.c

```diff
@@ -0,0 +1,845 @@
+		    (ARRAY_SIZE(msg_ram->tx_fifo) << CAN_MCAN_TXBC_TFQS_POS);
+	if (sizeof(msg_ram->tx_fifo[0].data) <= 24) {
+		can->txesc = (sizeof(msg_ram->tx_fifo[0].data) - 8) / 4;
+	} else {
+		can->txesc = (sizeof(msg_ram->tx_fifo[0].data) - 32) / 16 + 5;
+	}
+
+	if (sizeof(msg_ram->rx_fifo0[0].data) <= 24) {
+		can->rxesc = (((sizeof(msg_ram->rx_fifo0[0].data) - 8) / 4) <<
+				CAN_MCAN_RXESC_F0DS_POS) |
+			     (((sizeof(msg_ram->rx_fifo1[0].data) - 8) / 4) <<
+				CAN_MCAN_RXESC_F1DS_POS) |
+			     (((sizeof(msg_ram->rx_buffer[0].data) - 8) / 4) <<
+				CAN_MCAN_RXESC_RBDS_POS);
+	} else {
+		can->rxesc = (((sizeof(msg_ram->rx_fifo0[0].data) - 32)
+				/ 16 + 5) << CAN_MCAN_RXESC_F0DS_POS) |
+			     (((sizeof(msg_ram->rx_fifo1[0].data) - 32)
+				/ 16 + 5) << CAN_MCAN_RXESC_F1DS_POS) |
+			     (((sizeof(msg_ram->rx_buffer[0].data) - 32)
+				/ 16 + 5) << CAN_MCAN_RXESC_RBDS_POS);
+	}
+#endif
+
+#ifdef CONFIG_CAN_FD_MODE
+	can->cccr |= CAN_MCAN_CCCR_FDOE | CAN_MCAN_CCCR_BRSE;
+#else
+	can->cccr &= ~(CAN_MCAN_CCCR_FDOE | CAN_MCAN_CCCR_BRSE);
+#endif
+	can->cccr &= ~(CAN_MCAN_CCCR_TEST | CAN_MCAN_CCCR_MON |
+		       CAN_MCAN_CCCR_ASM);
+	can->test &= ~(CAN_MCAN_TEST_LBCK);
+
+#if defined(CONFIG_CAN_DELAY_COMP) && defined(CONFIG_CAN_FD_MODE)
+	can->dbtp |= CAN_MCAN_DBTP_TDC;
+#endif
+
+#ifdef CONFIG_CAN_STM32FD
+	can->rxgfc |= (CONFIG_CAN_MAX_STD_ID_FILTER << CAN_MCAN_RXGFC_LSS_POS) |
+		      (CONFIG_CAN_MAX_EXT_ID_FILTER << CAN_MCAN_RXGFC_LSE_POS) |
+		      (0x2 << CAN_MCAN_RXGFC_ANFS_POS) |
+		      (0x2 << CAN_MCAN_RXGFC_ANFE_POS);
+#else
+	can->gfc |= (0x2 << CAN_MCAN_GFC_ANFE_POS) |
+		    (0x2 << CAN_MCAN_GFC_ANFS_POS);
+#endif /* CONFIG_CAN_STM32FD */
+
+	timing.sjw = cfg->sjw;
+	if (cfg->sample_point) {
+		ret = can_calc_timing(dev, &timing, cfg->bus_speed,
+				      cfg->sample_point);
+		if (ret == -EINVAL) {
+			LOG_ERR("Can't find timing for given param");
+			return -EIO;
+		}
+		LOG_DBG("Presc: %d, TS1: %d, TS2: %d",
+			timing.prescaler, timing.phase_seg1, timing.phase_seg2);
+		LOG_DBG("Sample-point err : %d", ret);
+	} else if (cfg->prop_ts1) {
+		timing.prop_seg = 0;
+		timing.phase_seg1 = cfg->prop_ts1;
+		timing.phase_seg2 = cfg->ts2;
+		ret = can_calc_prescaler(dev, &timing, cfg->bus_speed);
+		if (ret) {
+			LOG_WRN("Bitrate error: %d", ret);
+		}
+	}
+#ifdef CONFIG_CAN_FD_MODE
+	timing_data.sjw = cfg->sjw;
+	if (cfg->sample_point_data) {
+		ret = can_calc_timing_data(dev, &timing_data,
+					   cfg->bus_speed_data,
+					   cfg->sample_point_data);
+		if (ret == -EINVAL) {
+			LOG_ERR("Can't find timing for given dataphase param");
+			return -EIO;
+		}
+
+		LOG_DBG("Sample-point err data phase: %d", ret);
+	} else if (cfg->prop_ts1_data) {
+		timing_data.prop_seg = 0;
+		timing_data.phase_seg1 = cfg->prop_ts1;
+		timing_data.phase_seg2 = cfg->ts2;
+		ret = can_calc_prescaler(dev, &timing_data,
+					 cfg->bus_speed_data);
+		if (ret) {
+			LOG_WRN("Dataphase bitrate error: %d", ret);
+		}
+	}
+#endif
```

Comment:
This looks like it's still using some of the arbitration phase variables instead of data phase. As a side note, I noticed this since it was causing a crash in `can_calc_prescalar` since prescaler ended up with 0 resulting in a divide by zero exception.